src/main/java/com/microsoft/jenkins/servicefabric/ServiceFabricPublishStep.java

```diff
@@ -0,0 +1,418 @@
+        return new SynchronousNonBlockingStepExecution<Void>(context) {
+            @SuppressWarnings("ConstantConditions")
+            @Override
+            protected Void run() throws Exception {
+                StepContext context = getContext();
+                FilePath workspace = context.get(FilePath.class);
+                workspace.mkdirs();
+
+                runStep(context,
+                        context.get(Run.class),
+                        workspace,
+                        context.get(Launcher.class),
+                        context.get(TaskListener.class));
+                return null;
+            }
+        };
+    }
+
+    @Override
+    public DescriptorImpl getDescriptor() {
+        return (DescriptorImpl) super.getDescriptor();
+    }
+
+
+    public void runStep(@Nullable StepContext context,
+                        @Nonnull Run<?, ?> run,
+                        @Nonnull FilePath workspace,
+                        @Nonnull Launcher launcher,
+                        @Nonnull TaskListener listener) throws InterruptedException, IOException {
+        String buildId = AppInsightsUtils.hash(run.getUrl());
+        AzureServiceFabricPlugin.sendEvent("StartDeploy",
+                Constants.AI_RUN, buildId);
+
+        try {
+            doRunStep(context, run, workspace, launcher, listener);
+            AzureServiceFabricPlugin.sendEvent("Deployed", Constants.AI_RUN, buildId);
+        } catch (InterruptedException | IOException | RuntimeException e) {
+            AzureServiceFabricPlugin.sendEvent("DeployFailed",
+                    Constants.AI_RUN, buildId,
+                    "Message", e.getMessage());
+            throw e;
+        }
+    }
+
+    private void doRunStep(@Nullable StepContext context,
+                           @Nonnull Run<?, ?> run,
+                           @Nonnull FilePath workspace,
+                           @Nonnull Launcher launcher,
+                           @Nonnull TaskListener listener) throws InterruptedException, IOException {
+        String buildId = AppInsightsUtils.hash(run.getUrl());
+
+        String cfgType = getConfigureType();
+        if (Constants.CONFIGURE_TYPE_SELECT.equals(cfgType)) {
+            Azure azure = AzureHelper.buildClient(azureCredentialsId);
+            AzureServiceFabricPlugin.sendEvent("DeployAzure",
+                    Constants.AI_RUN, buildId,
+                    "Subscription", AppInsightsUtils.hash(azure.subscriptionId()),
+                    "ResourceGroup", AppInsightsUtils.hash(resourceGroup),
+                    "Cluster", AppInsightsUtils.hash(serviceFabric));
+
+            ServiceFabricCluster cluster = new ServiceFabricCluster(azure, resourceGroup, serviceFabric);
+            String managementEndpoint = cluster.getManagementEndpoint();
+            try {
+                URL url = new URL(managementEndpoint);
+                if ("https".equalsIgnoreCase(url.getProtocol())) {
```

Comment:
Is it possible to check this before we run this pipeline step ?. Like doing a validation as done by "doFillServiceFabricItems" or "doCheckApplicationName".

Reply:
The `doCheckXXX` method only receives the value being checked (in this case, the Service Fabric resource name), which is not enough for us to retrieve the Service Fabric details.

We can add a "Verify" button and implement customized verification logic which pulls all the required configuration fields, and and do deep verification if user clicks the button. We may check the following items in this case:

* whether the service fabric is secured, and if yes, do we have certificate & key configured
* can we connect to the service fabric cluster

This is not a trivial change, so I suggest we do that in a separate PR.

Reply:
Ok Sure.